### PR TITLE
Installation scripts

### LINKED ISSUE #78: Feature Request: Support for pre, post, preun, postun scripts
This package is very useful for building RPMs and works very well so far.

I'd like to see support for adding various scriptlets to the spec file, for example:
- pre (before installation)
- post (after installation)
- preun (before uninstallation)
- postun (after uninstallation)

In my project, I need to execute some additional commands after installation and before uninstallation, so this feature would help me keeping a clean package structure.

Thanks in advance :)

### PR DESCRIPTION
**Pull request by fcastilloec**
This fixes #78

### CODE REVIEW

File: src/installer.js [hunk 1]
```diff
@@ -100,11 +100,31 @@ const getDefaults = function (data) {
...
 }
 
+/**
+ * Read scripts from provided filename and add them to the options
+ */
+const generateScripts = function (options) {
+  const scriptNames = ['pre', 'post', 'preun', 'postun']
+
+  return Promise.all(_.map(options.scripts, (item, key) => {
+    if (_.includes(scriptNames, key)) {
```
Comment (malept) [inline src/installer.js]: We should add a TODO here to replace this with `Array.prototype.includes` when we get to Node 6 in this module.

File: test/installer.js [hunk 2]
```diff
@@ -113,4 +114,30 @@ describe('module', function () {
 
     it('generates a `.rpm` package', () => assertNonASARRpmExists(outputDir))
   })
+
+  describeInstaller(
+    'with an app with installation scripts as paths',
```
Comment (malept) [inline test/installer.js]: This implies that you can specify installation scripts as not-paths (inline?) somehow...

File: resources/spec.ejs [hunk 3]
```diff
@@ -27,8 +27,22 @@ mkdir -p %{buildroot}/usr/
...
 %post
+<% if (post) {%><%= post %>
+<% } else { %>
 update-desktop-database &> /dev/null
+<% } %>
 
 %postun
+<% if (postun) {%><%= postun %>
+<% } else { %>
 update-desktop-database &> /dev/null
```
Comment (malept) [inline resources/spec.ejs]: Might want to specify in the docs that they'll need to update the desktop file cache themselves if they specify `post`/`postun` scripts.

File: test/installer.js (hunk 2 above)
Comment (fcastilloec) [reply test/installer.js]: Yes, it does. Inline is not the preferred way, as scripts can be lengthy, that's why I did not document this option. Currently, users need to pass a `scripts` object with the paths, but they can also pass `options.pre`, `options.post`, etc. containing the script's body as a string and that would also work. This is a side effect on how templates work with lodash. We need to send the `options` object to the template to generate the specs file and that object has to contain the contents of the scripts already, so people could just add the contents of the scripts directly to the `options` object

File: resources/spec.ejs (hunk 3 above)
Comment (fcastilloec) [reply resources/spec.ejs]: Updating the cache is only necessary if the software is handling some MIME types, which is not always the case and hence this should not always be called. I was actually considering removing the line altogether. On `-debian` we never had scripts until recently and hence we never called this and we still don't call this function even if the user doesn't specify scripts. I'm leaning towards removing the line. @malept thoughts?

Comment (malept) [reply resources/spec.ejs]: Perhaps add the line only if `options.mimeType` is passed in?

Comment (fcastilloec) [reply resources/spec.ejs]: That leaves us with a special cases when the user passes it's own script and also has MIME types. Do we append the command to the user script or just use the user script?
I think it's better to delete the line, that way this module works the same as `-debian`. I'm going to go with your original suggestion, of putting a note under `options.mimeType` that if it's used, `update-desktop-database &> /dev/null` should be added as `post` and `postun` scripts

File: README.md [hunk 4]
```diff
@@ -369,6 +369,28 @@ Default: `[]`
...
 MIME types the application is able to open, used in the [`MimeType` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).
 
+If this option is specified, make sure to run `update-desktop-database &> /dev/null` as part of the `post` and `postun` scripts to refresh the cache database of MIME types.
+You can read more about refreshing the [MIME-type/application cache](https://fedoraproject.org/wiki/NewMIMESystem).
+
+#### options.scripts
+Type: `Object[String:String]`
+Default: `undefined`
+
+Path to installation scripts with their corresponding name. The files contents will be added to the spec file.
```
Comment (malept) [inline README.md]: Perhaps

```suggestion
Path to [installation scripts](https://docs.fedoraproject.org/en-US/packaging-guidelines/Scriptlets/) with their corresponding name. The files contents will be added to the spec file.
```

And then drop the "you can read more..." after the example.

File: README.md [hunk 5]
```diff
@@ -369,6 +369,28 @@ Default: `[]`
 
 MIME types the application is able to open, used in the [`MimeType` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).
 
+If this option is specified, make sure to run `update-desktop-database &> /dev/null` as part of the `post` and `postun` scripts to refresh the cache database of MIME types.
```
Comment (malept) [inline README.md]:
```suggestion
If this option is specified, make sure to run `update-desktop-database &> /dev/null` as part of the `post` and `postun` scripts to refresh the [cached database of MIME types](https://fedoraproject.org/wiki/NewMIMESystem).
```

Also drop the following sentence.

**Pull request merged**